### ISSUE TITLE
TF Output is not updated, when aws_servicecatalog_provisioned_product replaces a resource

### ISSUE THREAD

**Issue by micweg1** [hashicorp/terraform-provider-aws#26549]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v1.1.9
registry.terraform.io/hashicorp/aws: 4.28.0 (latest)

### Terraform Configuration Files

**Service Catalog Product Yaml File**
```yaml
AWSTemplateFormatVersion: 2010-09-09
Description: Sumit Initial CloudFormation script

Parameters:
  AwsLogicalAccountNameLowerCase:
    Description: Logical AWS account name e.g. kush
    Type: String

Resources:
################S3 Buckets################
  KushResourceBucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: !Join ['', [!Ref AwsLogicalAccountNameLowerCase, '-learning-1000-resouces']]
      BucketEncryption:
        ServerSideEncryptionConfiguration:
        - ServerSideEncryptionByDefault:
            SSEAlgorithm: AES256
      PublicAccessBlockConfiguration:
          BlockPublicAcls: True
          BlockPublicPolicy: True
          IgnorePublicAcls: True
          RestrictPublicBuckets: True

################S3 Bucket Policies################
  KushResourceBucketPolicy:
    Type: AWS::S3::BucketPolicy
    Properties:
      Bucket: !Ref KushResourceBucket
      PolicyDocument:
        Statement:
          - Sid: DenyInsecureConnections
            Effect: Deny
            Principal: '*'
            Action: '*'
            Resource: !Join ['', ['arn:aws:s3:::', !Ref AwsLogicalAccountNameLowerCase, '-learning-1000-resouces/*']]
            Condition:
              Bool:
                aws:SecureTransport: false
          - Sid: DenyIncorrectEncryptionHeader
            Effect: Deny
            Principal: '*'
            Action:
              - s3:PutObject
            Resource: !Join ['', ['arn:aws:s3:::', !Ref AwsLogicalAccountNameLowerCase, '-learning-1000-resouces/*']]
            Condition:
              StringNotEquals:
                s3:x-amz-server-side-encryption: AES256
          - Sid: DenyUnEncryptedObjectUploads
            Effect: Deny
            Principal: '*'
            Action:
              - s3:PutObject
            Resource: !Join ['', ['arn:aws:s3:::', !Ref AwsLogicalAccountNameLowerCase, '-learning-1000-resouces/*']]
            Condition:
              'Null':
                s3:x-amz-server-side-encryption: true
          - Sid: DenyUnwantedCannedACLs
            Effect: Deny
            Principal: '*'
            Action:
              - s3:PutObject
            Resource: !Join ['', ['arn:aws:s3:::', !Ref AwsLogicalAccountNameLowerCase, '-learning-1000-resouces/*']]
            Condition:
              StringEquals:
                s3:x-amz-acl:
                  - public-read
                  - public-read-write
                  - aws-exec-read
                  - authenticated-read
                  - bucket-owner-read
                  - log-delivery-write

Outputs:
  BucketName:
    Description: Ref to the Bucket Name
    Value: !Ref KushResourceBucket

  BucketArn:
    Description: Bucket Arn
    Value: !GetAtt KushResourceBucket.Arn`
```

### Terraform Configuration Files

```hcl
#provider.tf

provider "aws" {
  region = "eu-central-1"
}
```

```hcl
#variables.tf

variable "product_id" {
  description = "Enter Product_Id"
  type        = string
  default     = "prod-yyvg6iyzlz6de"
}

variable "provisioning_artifact_id" {
  description = "Enter provisioning_artifact_id"
  type        = string
  default     = "pa-atcnxryqztky6"
}

variable "aws_logical_account_name_lower_case" {
  description = "Enter aws_logical_account_name_lower_case"
  type        = string
  default     = "michael-v2"
}

variable "provisioned_product_name" {
  description = "Enter provisioned_product_name"
  type        = string
  default     = "Bucket_Creation_michael"
}
```

```hcl
#main.tf

# Provisioning Service Catalog Product
resource "aws_servicecatalog_provisioned_product" "s3_bucket" {
  // Product Name
  name         = var.provisioned_product_name

  # Service catalog parameters
  product_id                = var.product_id
  provisioning_artifact_id  = var.provisioning_artifact_id
  // Service catalog provisioning parameters

  provisioning_parameters {
    key   = "AwsLogicalAccountNameLowerCase"
    value = var.aws_logical_account_name_lower_case
  }

}

locals {
  depends_on = [
    aws_servicecatalog_provisioned_product.s3_bucket
  ]
  pp_obtained_outputs = {
    for output in aws_servicecatalog_provisioned_product.s3_bucket.outputs : output.key =>
    output.value if contains(
      [ "BucketName",
        "BucketArn"],
    output.key)
  }
}

output "BucketName" {
  value       = local.pp_obtained_outputs["BucketName"]
  description = "Outputs the Bucket Name"
}

output "BucketArn" {
  value       = local.pp_obtained_outputs["BucketArn"]
  description = "Outputs the Bucket BucketArn"
}

```

### Expected Behavior

The module creates a Service Catalog provisioned product, which creates an S3 Bucket. The Bucket name is taken from the variable "aws_logical_account_name_lower_case". The Provisioned Product then outputs the ARN and the Bucket Name. The outputs of the Service Catalog Provisioned Product are then fetched and outputted by the Terraform module (in main.tf).

If we change the desired Bucket name from the variable, the Provisioned Product should be updated, destroying the old bucket, creating a new one, updating the output and Terraform should fetch the new ARN and Bucket Name that are outputted by the Service Catalog Provisioned Product..

### Actual Behavior

On the first apply, the Bucket is created and the output is correct.
When the Bucket Name is changed by the variable in Terraform, the Provisoned Product is correctly updated: The Bucket is replaced and the Service Catalog Output is also updated. But the Terraform Output is not updated. It still reflects the old status (old Bucket Name and ARN).

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`
3. `terraform apply`

The Provisioned Product including the Bucket is created correctly. Terraform gives the right output.

5. `Change Bucket name in the variables.tf`

Now, we change the Bucket name by modifying the variable "aws_logical_account_name_lower_case" (e.g. from my-bucket-1 to my-bucket-2)

6. `terraform plan`

Terraform will show that it wants to update the Service Catalog Provisioned Product. It does not display that there might be possible changes on the output.

7. `terraform apply`

The Provisioned Product is updated correctly. The Bucket is replaced and the Output of the Service Catalog Provisioned Product is updated. Nevertheless, Terraform does not fetch the updated output and still keeps the old one (e.g. in case of the Bucket Name output, it would keep "my-bucket-1" and would not update to "my-bucket-2)

**Comment by justinretzolk** [comment]
Hey @micweg1 👋 Thank you for taking the time to raise this! In case whoever picks this up to look into it needs, is it possible to supply debug logs (redacted as needed) as well?

**Comment by micweg1** [comment]
Hello @justinretzolk,

thanks for the first response on this case.
Please find the Debug log attached:
[TF_LOG.log](https://github.com/hashicorp/terraform-provider-aws/files/9461214/TF_LOG.log)
.

**Comment by kush-sumit** [comment]
hi @justinretzolk : any updates on this one please?

**Comment by florian0410** [comment]
Hello,

Thank you for raising this issue, I got the exact same problem on my side.
After digging a bit, I found out the outputs were not updated correctly depending on the update I do.


Creating the product
```bash
Terraform will perform the following actions:

  # aws_servicecatalog_provisioned_product.test will be created
  + resource "aws_servicecatalog_provisioned_product" "test" {
      + accept_language                        = "en"
      + arn                                    = (known after apply)
      + cloudwatch_dashboard_names             = (known after apply)
      + created_time                           = (known after apply)
      + id                                     = (known after apply)
      + ignore_errors                          = false
      + last_provisioning_record_id            = (known after apply)
      + last_record_id                         = (known after apply)
      + last_successful_provisioning_record_id = (known after apply)
      + launch_role_arn                        = (known after apply)
      + name                                   = (known after apply)
      + outputs                                = (known after apply)
      + path_id                                = "lpv2-cdcjxjskb4mok"
      + product_id                             = "prod-qn5hposz6iuss"
      + provisioning_artifact_id               = "pa-g4ox7dmbphccq"
      + retain_physical_resources              = false
      + status                                 = (known after apply)
      + status_message                         = (known after apply)
      + tags_all                               = (known after apply)
      + type                                   = (known after apply)

      + provisioning_parameters {
          + key   = "value1"
          + value = "v11"
        }
      + provisioning_parameters {
          + key   = "value2"
          + value = "v22"
        }
      + provisioning_parameters {
          + key   = "value3"
          + value = "v33"
        }
    }

  # random_uuid.generate_uuid will be created
  + resource "random_uuid" "generate_uuid" {
      + id     = (known after apply)
      + result = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + provisioned-product_outputs = (known after apply)
```

After execution I can find my outputs as expected

```
provisioned-product_outputs = toset([
  {
    "description" = ""
    "key" = "value1Out"
    "value" = "v11"
  },
  {
    "description" = ""
    "key" = "value2Out"
    "value" = "v22"
  },
  {
    "description" = ""
    "key" = "value3Out"
    "value" = "v33"
  },
  {
    "description" = "The ARN of the launched Cloudformation Stack"
    "key" = "CloudformationStackARN"
    "value" = "ANONYMIZED"
  },
])
```
It is coresponding to the output on provisioned product also: 
<img width="998" alt="Capture d’écran 2023-01-20 à 19 39 19" src="https://user-images.githubusercontent.com/10270633/213780391-63d7422d-e8d9-45e3-8e9d-8e346262a3da.png">


Now If I replace the template of the product with another version **including a new output named value4Out** and apply the change again, I can see following:

```
  # aws_servicecatalog_provisioned_product.test will be updated in-place
  ~ resource "aws_servicecatalog_provisioned_product" "test" {
        id                                     = "pp-6klh7ydf2waje"
        name                                   = "test_product_terraform_issue-36d9ad90-f38c-ca6f-9db8-3b849a6d664a"
      ~ provisioning_artifact_id               = "pa-g4ox7dmbphccq" -> "pa-ycc4ovdom2vce"
        tags                                   = {}
        # (15 unchanged attributes hidden)

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

And after apply I still see the exact same outputs...

```
Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

Outputs:

provisioned-product_outputs = toset([
  {
    "description" = ""
    "key" = "value1Out"
    "value" = "v11"
  },
  {
    "description" = ""
    "key" = "value2Out"
    "value" = "v22"
  },
  {
    "description" = ""
    "key" = "value3Out"
    "value" = "v33"
  },
  {
    "description" = "The ARN of the launched Cloudformation Stack"
    "key" = "CloudformationStackARN"
    "value" = "ANONYMIZED"
  },
])
```
We cannot find the newly added output but when looking at Service Catalog it has been deployed as expected.

<img width="900" alt="Capture d’écran 2023-01-20 à 19 39 25" src="https://user-images.githubusercontent.com/10270633/213780557-261fd1b3-09fe-47f6-94a9-f3db6fd635b0.png">


The only workaround I found for that is to use the refresh command to force terraform to align its outputs to Service Catalog ones.

```
terraform refresh
```


**Reproduction**

- Deploy a provisioned product with at least one output using terraform
- Go to product on ServiceCatalog service -> Seelect your product -> create new version -> provide a cloudformation template with a new output 
- Remove the previous version to keep only one
- terraform apply
- Check terraform state or print outputs of aws_servicecatalog_provisioned_product to see issue

**Comment by florian0410** [comment]
The issue is fixed since release of AWS provide 4.56.0, I get now the following on my second update

```
terraform apply
random_uuid.generate_uuid: Refreshing state... [id=299a507b-9f8d-d4cd-3b30-687a2ea3e249]
data.aws_servicecatalog_launch_paths.fetch_launch_path: Reading...
data.aws_caller_identity.current: Reading...
data.aws_caller_identity.current: Read complete after 2s [id=859324646650]
data.aws_servicecatalog_launch_paths.fetch_launch_path: Read complete after 4s [id=prod-qn5hposz6iuss]
aws_servicecatalog_provisioned_product.test: Refreshing state... [id=pp-gvc6ldgzym7mo]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_servicecatalog_provisioned_product.test will be updated in-place
  ~ resource "aws_servicecatalog_provisioned_product" "test" {
        id                                     = "pp-gvc6ldgzym7mo"
        name                                   = "test_product_terraform_issue-299a507b-9f8d-d4cd-3b30-687a2ea3e249"
      ~ outputs                                = [
          - {
              - description = ""
              - key         = "value1Out"
              - value       = "v11"
            },
          - {
              - description = ""
              - key         = "value2Out"
              - value       = "v22"
            },
          - {
              - description = ""
              - key         = "value3Out"
              - value       = "v33"
            },
          - {
              - description = "The ARN of the launched Cloudformation Stack"
              - key         = "CloudformationStackARN"
              - value       = "ANONYMIZED"
            },
        ] -> (known after apply)
      ~ provisioning_artifact_id               = "pa-hzslyw2vxtm56" -> "pa-cumnbef3cvcia"
        tags                                   = {}
        # (14 unchanged attributes hidden)

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Changes to Outputs:
  ~ provisioned-product_outputs = [
      - {
          - description = ""
          - key         = "value1Out"
          - value       = "v11"
        },
      - {
          - description = ""
          - key         = "value2Out"
          - value       = "v22"
        },
      - {
          - description = ""
          - key         = "value3Out"
          - value       = "v33"
        },
      - {
          - description = "The ARN of the launched Cloudformation Stack"
          - key         = "CloudformationStackARN"
          - value       = "ANONYMIZED"
        },
    ] -> (known after apply)
```

After apply the outputs are correctly updated

```
Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

Outputs:

provisioned-product_outputs = toset([
  {
    "description" = ""
    "key" = "value1Out"
    "value" = "v11"
  },
  {
    "description" = ""
    "key" = "value2Out"
    "value" = "v22"
  },
  {
    "description" = ""
    "key" = "value3Out"
    "value" = "v33"
  },
  {
    "description" = ""
    "key" = "value4Out"
    "value" = "v33"
  },
  {
    "description" = "The ARN of the launched Cloudformation Stack"
    "key" = "CloudformationStackARN"
    "value" = "ANONYMIZED"
  },
])
```

This ticket can be closed @micweg1

**Comment by micweg1** [comment]
Thanks @florian0410 for the support and thanks to the developers fixing this issue.

As this a fix is provider in AWS provider version 4.56.0, I will close this issue.

Thanks all!
